### PR TITLE
[Skylark] Use POJOs instead of dynamic proxies.

### PR DESCRIPTION
**Pull request by ttsugriy**
Java uses dynamically generated proxy classes to access annotation properties
and their methods are ~7X slower than plain getters. According to async-profiler
50%+ of `convertArgumentList` method time is spent in dynamic proxy methods, so
optimizing their performance makes sense.

This also makes the model less anemic, since POJOs can actually provide business
methods.

### CODE REVIEW

**Comment by ttsugriy** [comment]
proof of proxy vs getter perf - https://gist.github.com/ttsugriy/c5c7719b9d7bb993d62f445a0be94e5c

**Comment by ttsugriy** [comment]
@laurentlb, @dslomov, would you mind taking a look? On top of having ~25% parse time improvement this change also enables further optimizations - I'll move the getType computation to a `ParamDescriptor.of` static factory method which should eliminate the next bottleneck - going through loading cache to resolve types, since types have to be resolved for every single method invocation and for each parameter.

File: src/main/java/com/google/devtools/build/lib/syntax/FuncallExpression.java [hunk 1]
```diff
@@ -57,29 +58,294 @@
...
+      return generic1;
+    }
+
+    static ParamTypeDescriptor of(ParamType paramType) {
+      return new ParamTypeDescriptor(paramType.type(), paramType.generic1());
+    }
+  }
+
+  /** A value class for storing {@link Param} metadata to avoid using Java proxies. */
+  public static final class ParamDescriptor {
```
Comment (c-parsons) [inline src/main/java/com/google/devtools/build/lib/syntax/FuncallExpression.java]: Could you please move this class to the top level? Long class definitions, even if they're trivial, make reading the containing class a bit more daunting.

File: src/main/java/com/google/devtools/build/lib/syntax/FuncallExpression.java [hunk 2]
```diff
@@ -57,29 +58,294 @@
...
+    /** @see Param#defaultValue() */
+    public String getDefaultValue() {
+      return defaultValue;
+    }
+  }
+
   /**
-   * A value class to store Methods with their corresponding SkylarkCallable annotations.
-   * This is needed because the annotation is sometimes in a superclass.
+   * A value class to store Methods with their corresponding SkylarkCallable annotations. This is
```
Comment (c-parsons) [inline src/main/java/com/google/devtools/build/lib/syntax/FuncallExpression.java]: Update javadoc

Does this now contain all of the information that was before stored in SkylarkCallable? Then we can remove that field

File: src/main/java/com/google/devtools/build/lib/syntax/SkylarkSignatureProcessor.java [hunk 3]
```diff
@@ -225,25 +225,31 @@ private static boolean isParamNamed(Param param) {
...
-                .setGlobals(Environment.CONSTANTS_ONLY)
-                .setEventHandler(Environment.FAIL_FAST_HANDLER)
-                .build()
-                .update("unbound", Runtime.UNBOUND);
-        return BuildFileAST.eval(env, param.defaultValue());
-      } catch (Exception e) {
-        throw new RuntimeException(String.format(
-            "Exception while processing @SkylarkSignature.Param %s, default value %s",
-            param.name(), param.defaultValue()), e);
+      if (defaultValue.isEmpty()) {
```
Comment (c-parsons) [inline src/main/java/com/google/devtools/build/lib/syntax/SkylarkSignatureProcessor.java]: Why the change here?

File: src/main/java/com/google/devtools/build/lib/syntax/FuncallExpression.java (hunk 1 above)
Comment (ttsugriy) [reply src/main/java/com/google/devtools/build/lib/syntax/FuncallExpression.java]: absolutely, I've used a nested class because MethodDescriptor is nested, but given how big all of these classes are, I agree that separating them makes code more readable :)

File: src/main/java/com/google/devtools/build/lib/syntax/FuncallExpression.java (hunk 2 above)
Comment (ttsugriy) [reply src/main/java/com/google/devtools/build/lib/syntax/FuncallExpression.java]: It does, but unfortunately `getAnnotation` is still used in `SkylarkSignatureProcessor` :(

File: src/main/java/com/google/devtools/build/lib/syntax/SkylarkSignatureProcessor.java (hunk 3 above)
Comment (ttsugriy) [reply src/main/java/com/google/devtools/build/lib/syntax/SkylarkSignatureProcessor.java]: it's so that this method can be used by clients that have `ParamDescriptor` or `Param` annotation. It's somewhat unfortunate since having parameters of the same type is a bit confusing, but until all usages of annotations are removed this is the way to avoid duplication :(

**Comment by laurentlb** [comment]
cc @meisterT

File: src/main/java/com/google/devtools/build/lib/syntax/SkylarkSignatureProcessor.java (hunk 3 above)
Comment (ttsugriy) [reply src/main/java/com/google/devtools/build/lib/syntax/SkylarkSignatureProcessor.java]: no problem at all, @c-parsons! Thanks for working on these changes! The only explanation I can think of with regards to your proxy overhead observation is that in the code you have used for benchmarking you didn't have a lot of parameters/methods. Since Buck does not have extension support, we end up having extremely complicated macros with hundreds (200+) parameters, so for us it ends up using a lot of cycles :( I'll gladly rebase my changes. Thanks again for working on Skylark perf and your reviews!

Comment (ttsugriy) [reply src/main/java/com/google/devtools/build/lib/syntax/SkylarkSignatureProcessor.java]: btw, @c-parsons, what do you think about moving the parsed `defaultValue` to the `ParamDescriptor`? At the moment a lot of time is spent in a `LoadingCache`, so storing the variable directly next to a parameter can make for a potentially easier to use API and better for performance.

**Comment by ttsugriy** [comment]
@c-parsons, I've effectively merged the `getType` caching from my other PR into this one, so please take a look at this one and I'll close the other one :)

File: src/main/java/com/google/devtools/build/lib/syntax/ParamDescriptor.java [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+    this.defaultValue = defaultValue;
+    this.type = type;
+    this.allowedTypes = allowedTypes;
+    this.generic1 = generic1;
+    this.callbackEnabled = callbackEnabled;
+    this.noneable = noneable;
+    this.named = named;
+    this.legacyNamed = legacyNamed;
+    this.positional = positional;
+
```
Comment (c-parsons) [inline src/main/java/com/google/devtools/build/lib/syntax/ParamDescriptor.java]: nit: kill extra newline

File: src/main/java/com/google/devtools/build/lib/syntax/MethodDescriptor.java [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+import com.google.devtools.build.lib.skylarkinterface.SkylarkCallable;
+import java.lang.reflect.Method;
+import java.util.Arrays;
+
+/**
+ * A value class to store Methods with their corresponding {@link SkylarkCallable} annotation
+ * metadata. This is needed because the annotation is sometimes in a superclass.
+ *
+ * <p>The annotation metadata is duplicated in this class to avoid usage of Java dynamic proxies
+ * which are ~7X slower
```
Comment (c-parsons) [inline src/main/java/com/google/devtools/build/lib/syntax/MethodDescriptor.java]: nit: period at end of javadoc

**Comment by ttsugriy** [comment]
Thanks for the supersonic review, @c-parsons! I've addressed the nits, so please take another look when you have a chance :)

**Comment by laurentlb** [comment]
I made small changes when importing your PR to fix lint issues (copyright header, unused fields, functions with same name should be grouped).

**Comment by ttsugriy** [comment]
Thank you, @laurentlb!
